PetAdoptions/petlistadoptions-py/app.py

```diff
@@ -0,0 +1,235 @@
+                ]
+            )
+            
+            for param in response['Parameters']:
+                if param['Name'] == '/petstore/rdssecretarn':
+                    self.rds_secret_arn = param['Value']
+                elif param['Name'] == '/petstore/searchapiurl':
+                    self.pet_search_url = param['Value']
+                    
+        except Exception as e:
+            logger.error(f"Failed to fetch from Parameter Store: {e}")
+    
+    def _get_database_connection_string(self) -> str:
+        """Get database connection string from AWS Secrets Manager"""
+        try:
+            # Check if this is a local test setup
+            if self.rds_secret_arn == "local-secret":
+                # Read from local file for testing
+                with open("/app/local-secret.json", "r") as f:
+                    secret_data = json.loads(f.read())
+            else:
+                # Use AWS Secrets Manager
+                secrets = boto3.client('secretsmanager')
+                response = secrets.get_secret_value(SecretId=self.rds_secret_arn)
+                secret_data = json.loads(response['SecretString'])
+            
+            # Build connection string for psycopg2
+            connection_string = f"postgresql://{secret_data['username']}:{secret_data['password']}@{secret_data['host']}:{secret_data.get('port', 5432)}/{secret_data['dbname']}"
+            
+            logger.info(f"Generated connection string for host: {secret_data['host']}")
+            return connection_string
+            
+        except Exception as e:
+            logger.error(f"Failed to get database connection string: {e}")
+            raise
+    
+    @contextmanager
+    def _get_database_connection(self):
+        """Context manager for database connections"""
+        connection_string = self._get_database_connection_string()
+        conn = psycopg2.connect(connection_string)
+        try:
+            yield conn
+        finally:
+            conn.close()
+    
+    def _get_latest_adoptions(self) -> List[Dict[str, Any]]:
+        """Get latest adoptions from database"""
+        with self._get_database_connection() as conn:
+            with conn.cursor() as cursor:
+                cursor.execute("SELECT pet_id, transaction_id, adoption_date FROM transactions ORDER BY id DESC LIMIT 25")
+                rows = cursor.fetchall()
+                
+                return [
+                    {
+                        'pet_id': pet_id,
+                        'transaction_id': transaction_id,
+                        'adoption_date': adoption_date.isoformat() if adoption_date else None
+                    }
+                    for pet_id, transaction_id, adoption_date in rows
+                ]
+    
+    def _search_pet_info(self, pet_id: str) -> List[Dict[str, Any]]:
+        """Search for pet information by pet_id"""
+        url = f"{self.pet_search_url}petid={pet_id}"
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Log Injection occurs when unsafe input is directly written to log files without proper sanitization. This can allow attackers to manipulate log entries, potentially leading to security issues like log forging or cross-site scripting. To prevent this, always sanitize user input before logging by removing or encoding newline characters, using string encoding functions, and leveraging built-in sanitization features of logging libraries when available. Learn more - https://cwe.mitre.org/data/definitions/117.html